src/bookshelf-sender/lib/bookshelf/sender/rsync_worker.rb

```diff
@@ -0,0 +1,130 @@
+    # RsyncWorker
+    #
+    # This class is reponsible for running rsync in response
+    # to updates from a queue.
+    #
+    # Updates are expected to be either a PendingUpdateSet or the
+    # symbol :full_update.  When reading updates, we will attempt to
+    # clear the entire queue to avoid uncessary calls to rsync.
+    #
+    # When :full_update is received, we will clear any pending updates
+    # from the queue and run a full sync. If the full_update fails, we
+    # will requeue the :full_update signal after waiting
+    # Configuration[:retry_interval] seconds.
+    #
+    # When a PendingUpdateSet fails we will requeue the
+    # PendingUpdateSet after waiting Configuration[:retry_interval]
+    # seconds.
+    #
+    class RsyncWorker
+      RSYNC_EXE = "rsync"
+      RSYNC_ARGS = "-r --delete -vvvvvv"
+
+      attr_reader :queue, :remote, :directory
+
+      def initialize(remote, directory, queue)
+        @remote = remote
+        @directory = directory
+        @queue = queue
+      end
+
+      def run
+        Thread.new do
+          Log.info("Starting RsyncWorker")
+          internal_run
+        end
+      end
+
+      private
+
+      def internal_run
+        loop do
+          updates = updates_from_queue # blocks until there are updates
+          Log.debug "Updates seen: #{updates.inspect}"
+          case updates
+          when PendingUpdateSet
+            do_incremental_update(updates)
+          when :full_sync
+            do_full_sync
+          else
+            Log.warn("Unknown update on queue: #{update.inspect}; Ignoring")
+          end
+        end
+      end
+
+      def do_full_sync
+        Log.info("Starting full sync.")
+        Log.info("Clearing remaining messages from queue.")
+        queue.clear
+        if run_rsync
+          Log.info("Full update succeeded.")
+        else
+          Log.info("Full update failed.")
+          retry_message(:full_sync)
+        end
+      end
+
+      def do_incremental_update(updates)
+        Tempfile.create("bookshelf-sender-incr-updates") do |f|
+          f.puts(updates.to_a)
+          f.flush
+          if run_rsync("--files-from #{f.path}")
+            Log.info("Incremental update succeeded.")
+          else
+            Log.info("Incremental update failed.")
+            retry_message(updates)
+          end
+        end
+      end
+
+      # Collect anything remaining on the queue.
+      #
+      # The goal of this function to help prevent running Rsync
+      # multiple times when we can avoid it.
+      def updates_from_queue
+        updates = PendingUpdateSet.new
+        loop do
+          an_updates = queue.pop
+
+          case an_updates
+          when :full_sync
+            # If our message is full_sync don't even bother, we are going
+            # to clear the queue anyway.
+            return :full_sync
+          when PendingUpdateSet
+            updates += an_updates
+          end
+
+          return updates if queue.empty?
```

Comment:
I think we might want to remove this attempt at getting multiple items on the queue or add some sort of upper bound.  I'm worried that with a fast enough writer/the right thread scheduling, this will never return.  The UpdateWriter already does work to minimize the items it puts on the queue.